### ISSUE TITLE
AIO does not work with BeagleBone Black / Suggested Improvements

### ISSUE THREAD

**Issue by nrcrast** [intel-iot-devkit/mraa#749]
You have some math in aio.c:
`pin = aio + board->gpio_count;`

The intention is that aio pins are always defined after regular GPIOs, which I find to be an odd generalization. This does not work with the beaglebone black. The gpio_count is currently 45, but looking at the pinout the AIN pins start at 81. 

Furthermore, the gpio_count is *only* used for the aio stuff. I think it would make more sense to define the aio pins as devices similar to the way uart, spi, and i2c work. Each board would have an array of aio devices that map to the pins:
`board->aio[0].pin = 81`

This would push the logic back down into the board specific modules where I think it belongs anyway.

**Comment by nrcrast** [comment]
Nevermind, did a little more reading, and it seems the beaglebone just needs some more work.

**Comment by nrcrast** [comment]
Sorry guys, :). 

Yeah, so part of the problem can be alleviated by defining aio.pinmap for the aio capable GPIOs on the beaglebone, but the problem is that the AIO pins on the beaglebone are not sequential. For example:

P9_33 -> AIN4
P9_34 -> GNDA_ADC
P9_35 -> AIN5

As I said before, I think more logic needs to be pushed down to the board level.

**Comment by nrcrast** [comment]
I have a proposed fix, I will make a new branch for it (separate from my other beaglebone stuff) and submit a PR for it.

**Comment by arfoll** [comment]
Yes, you're absolutely right our AIO module is lacking a bit of love, this is mostly because most/all? our boards with onboard ADCs use an arduino style pinout and we get away with it with a mixture of hacks but they won't work for BBB.

My suggestion for a quick fix would be to just check for an arduino style pinout (makes a few assumptions):
```
    if (board->gpio_count == 14 && board->phy_pin_count == 20) {
        // detecting arduino style board so adding 14 to pincounts
        pin = aio + board->gpio_count;
    }
```

I'd have to double check this won't break other boards. But as a starting point we could have this.

As I typed this you said you had a proposed fix so happy to have a look at that!

**Comment by Propanu** [comment]
I've been looking into this too, while this worked well for the first couple of boards we added to MRAA, there are several new ones out there that need this workaround.

**Comment by nrcrast** [comment]
@Propanu proposed patch here #750
